### PR TITLE
CSRF token storage by entity type

### LINKED ISSUE #5827: Still getting '757: unexpected token at "Bad Request"'
### New Issue Checklist
- [X] Updated fastlane to the latest version

### Issue Description
I was expecting the fix during the weekend to have solved it, but that was not the case - so I have created a new issue, even though it seems to be like the other issues it might not then.
I run produce with fastlane and gets this error.

##### Complete output when running fastlane, including the stack trace and command used

```
+----------------+---------------------------------------+
|               Summary for produce 1.2.0                |
+----------------+---------------------------------------+
| username       | mho@vertica.dk                        |
| app_identifier | dk.vertica.daka.debug                 |
| language       | Danish                                |
| app_version    | 1.1.0                                 |
| app_name       | PigUp & Ko dev                        |
| sku            | dk.vertica.daka.development.pigupogko |
| skip_itc       | true                                  |
| skip_devcenter | false                                 |
| team_name      | Vertica A/S                           |
| team_id        | QGARR49P23                            |
| itc_team_id    | QGARR49P23                            |
+----------------+---------------------------------------+

[08:49:58]: Creating new app 'PigUp  Ko dev' on the Apple Dev Center
[08:50:00]: -----------------
[08:50:00]: --- Step: say ---
[08:50:00]: -----------------
[08:50:00]: $ say '757: unexpected token at "Bad Request"'
[08:50:05]: Variable Dump:
[08:50:05]: {:PLATFORM_NAME=>nil, :LANE_NAME=>"create_app_for_dev_and_test", :VERSION_NUMBER=>"1.1.0", :CERT_FILE_PATH=>"/Users/mortenholmgaard/Documents/Code/daka-ios/fastlane/certificates/LU924LTGCW.cer", :CERT_CERTIFICATE_ID=>"LU924LTGCW"}
[08:50:05]: 757: unexpected token at 'Bad Request'

+------+-------------------------------------+-------------+
|                     fastlane summary                     |
+------+-------------------------------------+-------------+
| Step | Action                              | Time (in s) |
+------+-------------------------------------+-------------+
| 1    | Verifying required fastlane version | 0           |
| 2    | get_version_number                  | 0           |
| 3    | team_id                             | 0           |
| 4    | prompt                              | 0           |
| 5    | cert                                | 8           |
| 6    | produce                             | 5           |
| 7    | say                                 | 4           |
+------+-------------------------------------+-------------+

[08:50:05]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...

➡️  "unexpected token at 'Bad Request'" from Spaceship when attempting to repair a provisioning profile
   https://github.com/fastlane/fastlane/issues/5813 [open] 23 💬
   40 minutes ago

➡️  Bad Request - portal_client.rb - associate_groups_with_app
   https://github.com/fastlane/fastlane/issues/2049 [open] 0 💬
   07 Mar 2016

➡️  757: unexpected token at 'HTTP 400 Bad Request'
   https://github.com/fastlane/fastlane/issues/2947 [closed] 4 💬
   08 Mar 2016

and 20 more at: https://github.com/fastlane/fastlane/search?q=757:%20unexpected%20token%20at%20'Bad%20Request'&type=Issues&utf8=✓
/Library/Ruby/Gems/2.0.0/gems/json-1.8.3/lib/json/common.rb:155:in `parse': [!] 757: unexpected token at 'Bad Request' (Faraday::ParsingError)
	from /Library/Ruby/Gems/2.0.0/gems/json-1.8.3/lib/json/common.rb:155:in `parse'
	from /Library/Ruby/Gems/2.0.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response/parse_json.rb:11:in `block in <class:ParseJson>'
	from /Library/Ruby/Gems/2.0.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:48:in `call'
	from /Library/Ruby/Gems/2.0.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:48:in `parse'
	from /Library/Ruby/Gems/2.0.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:39:in `process_response'
	from /Library/Ruby/Gems/2.0.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:32:in `block in call'
	from /Library/Ruby/Gems/2.0.0/gems/faraday-0.9.2/lib/faraday/response.rb:57:in `on_complete'
	from /Library/Ruby/Gems/2.0.0/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response_middleware.rb:30:in `call'
	from /Library/Ruby/Gems/2.0.0/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:139:in `build_response'
	from /Library/Ruby/Gems/2.0.0/gems/faraday-0.9.2/lib/faraday/connection.rb:377:in `run_request'
	from /Library/Ruby/Gems/2.0.0/gems/faraday-0.9.2/lib/faraday/connection.rb:177:in `post'
	from /Users/mortenholmgaard/.gems/gems/spaceship-0.31.6/lib/spaceship/client.rb:411:in `block in send_request'
	from /Users/mortenholmgaard/.gems/gems/spaceship-0.31.6/lib/spaceship/client.rb:328:in `with_retry'
	from /Users/mortenholmgaard/.gems/gems/spaceship-0.31.6/lib/spaceship/client.rb:410:in `send_request'
	from /Users/mortenholmgaard/.gems/gems/spaceship-0.31.6/lib/spaceship/client.rb:385:in `request'
	from /Users/mortenholmgaard/.gems/gems/spaceship-0.31.6/lib/spaceship/portal/portal_client.rb:178:in `create_app!'
	from /Users/mortenholmgaard/.gems/gems/spaceship-0.31.6/lib/spaceship/portal/app.rb:97:in `create!'
	from /Users/mortenholmgaard/.gems/gems/produce-1.2.0/lib/produce/developer_center.rb:22:in `create_new_app'
	from /Users/mortenholmgaard/.gems/gems/produce-1.2.0/lib/produce/developer_center.rb:8:in `run'
	from /Users/mortenholmgaard/.gems/gems/produce-1.2.0/lib/produce/manager.rb:7:in `start_producing'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/actions/produce.rb:19:in `block in run'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/actions/produce.rb:17:in `chdir'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/actions/produce.rb:17:in `run'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:190:in `block (2 levels) in execute_action'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/actions/actions_helper.rb:35:in `execute_action'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:175:in `block in execute_action'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:174:in `chdir'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:174:in `execute_action'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:100:in `trigger_action_by_name'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/fast_file.rb:140:in `method_missing'
	from Fastfile:231:in `produce_app'
	from Fastfile:227:in `produce_devcenter'
	from Fastfile:206:in `setup_for_dev'
	from Fastfile:29:in `block in parsing_binding'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/lane.rb:33:in `call'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/lane.rb:33:in `call'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:49:in `block in execute'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:45:in `chdir'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/runner.rb:45:in `execute'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/lane_manager.rb:46:in `cruise_lane'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/command_line_handler.rb:30:in `handle'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/commands_generator.rb:49:in `block (2 levels) in run'
	from /Users/mortenholmgaard/.gems/gems/commander-4.4.0/lib/commander/command.rb:178:in `call'
	from /Users/mortenholmgaard/.gems/gems/commander-4.4.0/lib/commander/command.rb:178:in `call'
	from /Users/mortenholmgaard/.gems/gems/commander-4.4.0/lib/commander/command.rb:153:in `run'
	from /Users/mortenholmgaard/.gems/gems/commander-4.4.0/lib/commander/runner.rb:444:in `run_active_command'
	from /Users/mortenholmgaard/.gems/gems/fastlane_core-0.50.2/lib/fastlane_core/ui/fastlane_runner.rb:36:in `run!'
	from /Users/mortenholmgaard/.gems/gems/commander-4.4.0/lib/commander/delegates.rb:15:in `run!'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/commands_generator.rb:246:in `run'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/commands_generator.rb:20:in `start'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/lib/fastlane/cli_tools_distributor.rb:58:in `take_off'
	from /Users/mortenholmgaard/.gems/gems/fastlane-1.101.0/bin/fastlane:5:in `<top (required)>'
	from /usr/local/bin/fastlane:23:in `load'
	from /usr/local/bin/fastlane:23:in `<main>'
```

### Configuration Files

Please copy the complete content of your `Fastfile` and any other configuration files you use below:

Fastfile:
```ruby
  produce(
    username: ENV["APPLE_ID"],
    app_identifier: ENV["APP_IDENTIFIER"],
    language: "Danish",
    app_version: ENV["APP_VERSION_NUMBER"],
    app_name: ENV["APP_TITLE"],
    sku: "dk.vertica.daka.development.pigupogko",
    skip_itc: !create_itc,
    skip_devcenter: !create_developer_account,
    team_name: "teamname")  
```

### Environment

fastlane version (run `fastlane -v`):
fastlane 1.101.0
Do you use bundler to execute fastlane (i.e. `bundle exec fastlane`)?
No
Do you use a Ruby environment manager (e.g. `chruby`, `rbenv`, `rvm`)?
Do not think so

### PR DESCRIPTION
**Pull request by asfalcone**
Fixes #5827

### CODE REVIEW

File: spaceship/lib/spaceship/client.rb [hunk 1]
```diff
@@ -356,20 +356,41 @@ def do_login(user, password)
...
     # Is called from `parse_response` to store the latest csrf_token (if available)
-    def store_csrf_tokens(response)
+    def store_csrf_tokens(response, expected_key)
       if response and response.headers
         tokens = response.headers.select { |k, v| %w(csrf csrf_ts).include?(k) }
         if tokens and !tokens.empty?
-          @csrf_tokens = tokens
+         # TODO replace this by storing the expected_key in the hash directly
+
+          key = case expected_key
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/client.rb]: Why not just use `expected_key` as key here?

Comment (asfalcone) [inline spaceship/lib/spaceship/client.rb]: :) https://github.com/fastlane/fastlane/pull/5872/files#diff-a9086da2d7e8a66b647fdabdb0a7cca7R363

File: spaceship/lib/spaceship/portal/portal_client.rb [hunk 2]
```diff
@@ -407,27 +416,34 @@ def repair_provisioning_profile!(profile_id, name, distribution_method, app_id,
...
       # I couldn't find out why, the first response does have a valid Set-Cookie header
       # But it still needs this second request
-      App.all
+      klass.all
 
-      @requested_csrf_token = true if csrf_tokens.count > 0
+      csrf_cache[klass] = self.csrf_tokens
     end
   end
+  # rubocop:enable Metrics/ClassLength
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/portal/portal_client.rb]: We should remove this rule in `.rubocop.yml`, maybe in a separate PR though?

**Comment by KrauseFx** [comment]
This is great, ready to be merged 👍

File: spaceship/lib/spaceship/portal/portal_client.rb [hunk 3]
```diff
@@ -142,7 +143,7 @@ def update_service_for_app(app, service)
     end
 
     def associate_groups_with_app(app, groups)
-      ensure_csrf
+      ensure_csrf(Spaceship::AppGroup)
```
Comment (mfurtak) [inline spaceship/lib/spaceship/portal/portal_client.rb]: Did you try using the unqualified class name here? Or is the FQN just preferable?

Comment (asfalcone) [inline spaceship/lib/spaceship/portal/portal_client.rb]: Unqualified will work, but it seemed more clear to me?

**Comment by mfurtak** [comment]
Now that we recognize that the correct CSRF token must be used with the correct endpoint, I just wanted to point out that having the shared `csrf_tokens` value change in-between method calls feels like a dangerous thing. 

I don't know what kind of thread safety expectations people have for `spaceship`, though!

**Pull request merged**